subprojects/plugins/src/main/java/org/gradle/api/plugins/JavaPluginExtension.java

```diff
@@ -155,4 +159,106 @@
     @Incubating
     void consistentResolution(Action<? super JavaResolutionConsistency> action);
 
+    /**
+     * Configures the source sets of this project.
+     *
+     * <p>The given closure is executed to configure the {@link SourceSetContainer}. The {@link SourceSetContainer}
+     * is passed to the closure as its delegate.
+     * <p>
+     * See the example below how {@link org.gradle.api.tasks.SourceSet} 'main' is accessed and how the {@link org.gradle.api.file.SourceDirectorySet} 'java'
+     * is configured to exclude some package from compilation.
+     *
+     * <pre class='autoTested'>
+     * plugins {
+     *     id 'java'
+     * }
+     *
+     * sourceSets {
+     *   main {
+     *     java {
+     *       exclude 'some/unwanted/package/**'
+     *     }
+     *   }
+     * }
+     * </pre>
+     *
+     * @param closure The closure to execute.
+     * @return NamedDomainObjectContainer&lt;org.gradle.api.tasks.SourceSet&gt;
+     * @since 7.1
+     */
+    Object sourceSets(Closure closure);
```

Comment:
Would making this method take an Action instead break things?

Reply:
Ah, I see now.

Reply:
My intention here is to move the API around and not to do any major cleanup (except for simple `Property` conversions). Yes, it would make sense to do those simplifications, but I think that belongs to a different story.